File: leetcode/821.go
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func shortestToChar(s string, c byte) []int {
	var res []int
	cindex := -1
	for idx, x := range s {
		if x == rune(c) {
			cindex = idx
			break
		}
	}
	for idx, x := range s {
		if x == rune(c) {
			cindex = idx
		}
		res = append(res, abs(idx-cindex))
	}
	for idx := len(s) - 1; idx >= 0; idx-- {
		x := abs(idx - cindex)
		if res[idx] > x {
			res[idx] = x
		}
		if s[idx] == c {
			cindex = idx
		}
	}
	return res
}

File: leetcode/821.cpp
class Solution {
public:
    vector<int> shortestToChar(string s, char c) {
        vector<int> res;
        int cindex = -1;
        for(int i=0; i<s.size(); ++i) {
            if(s[i]==c) {
                cindex=i;
                break;
            }
        }
        for(int i=0; i<s.size(); ++i) {
            if(s[i]==c) cindex=i;
            res.push_back(abs(i-cindex));
        }
        for(int i=s.size()-1; i>=0; --i) {
            int x = abs(i-cindex);
            res[i]=min(x, res[i]);
            if(s[i]==c) cindex=i;
        }
        return res;
    }
};